File: test/typescript/steps/user_steps.ts
import { Given, When, Then } from '@cucumber/cucumber';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'path';
import { authContext } from './auth_steps';

interface FormSubmission {
  id?: string;
  form_type?: string;
  form_instance?: string;
  email?: string;
  ttc_option?: string;
  status?: string;
  data?: Record<string, unknown>;
  form_instance_page_data?: Record<string, unknown>;
}

interface StoredFormData {
  data: Record<string, unknown>;
  form_instance_page_data: Record<string, unknown>;
  form_instance_display: string;
  is_agreement_accepted: boolean;
  is_form_submitted: boolean;
  is_form_complete: boolean;
  last_update_datetime: string;
}

class MockTTCPortalUser {
  public email: string | null = null;
  public formData: Record<string, Record<string, StoredFormData>> = {};
  public config: Record<string, unknown> = {};
  public isProfileComplete: Record<string, boolean> = {};

  initializeUser(userDict: { email?: string }): void {
    this.email = userDict.email ?? null;
    this.formData = {};
    this.config = {};
    this.isProfileComplete = {};
  }

  setFormData(
    formType: string,
    formInstance: string,
    formData: Record<string, unknown>,
    formInstancePageData: Record<string, unknown>,
    formInstanceDisplay: string
  ): void {
    if (!this.formData[formType]) {
      this.formData[formType] = {};
    }

    // Compute is_form_complete by checking if all fields have non-empty values
    const isFormComplete = Object.values(formData).every(
      (v) => v !== null && v !== undefined && v !== ''
    );

    this.formData[formType][formInstance] = {
      data: formData,
      form_instance_page_data: formInstancePageData,
      form_instance_display: formInstanceDisplay,
      is_agreement_accepted: false,
      is_form_submitted: false,
      is_form_complete: isFormComplete,
      last_update_datetime: '2024-01-01 00:00:00',
    };

    // Also write to 'default' instance if this is a non-default instance
    if (formInstance !== 'default') {
      this.formData[formType]['default'] = this.formData[formType][formInstance];
    }
  }

  getFormData(formType: string, formInstance: string): StoredFormData | null {
    if (this.formData[formType] && this.formData[formType][formInstance]) {
      return this.formData[formType][formInstance];
    }
    return null;
  }

  loadUserData(userEmail: string): void {
    this.email = userEmail;
    this.initializeUser({ email: userEmail });
  }

  saveUserData(): void {
    // Mock save - no-op
  }

  getConfig(): Record<string, unknown> {
    return this.config;
  }

  setConfig(configParams: Record<string, unknown> | string): void {
    let params: Record<string, unknown>;

    if (typeof configParams === 'string') {
      params = JSON.parse(configParams);
    } else {
      params = configParams;
    }

    // Update config with provided params
    for (const [key, value] of Object.entries(params)) {
      this.config[key] = value;
    }
  }

  getFormInstances(
    formType: string
  ): Record<string, { page_data: Record<string, unknown>; display: string }> {
    /** Get all form instances for a form type, excluding 'default'. */
    const formInstances: Record<string, { page_data: Record<string, unknown>; display: string }> =
      {};

    if (this.formData[formType]) {
      for (const instanceId in this.formData[formType]) {
        if (instanceId !== 'default') {
          const instance = this.formData[formType][instanceId];
          formInstances[instanceId] = {
            page_data: instance.form_instance_page_data,
            display: instance.form_instance_display,
          };
        }
      }
    }

    return formInstances;
  }
}

interface UserFormContext {
  lastUpload?: {
    formType: string;
    formInstance: string;
    formData: Record<string, unknown>;
    formInstanceDisplay: string;
    formInstancePageData: Record<string, unknown>;
    user: MockTTCPortalUser;
  };
  user?: MockTTCPortalUser;
  formType?: string;
  formInstances?: Record<string, { page_data: Record<string, unknown>; display: string }>;
}

const userFormContext: UserFormContext = {};

let cachedSubmissions: FormSubmission[] | null = null;

function loadFormSubmissions(): FormSubmission[] {
  if (cachedSubmissions) {
    return cachedSubmissions;
  }
  const submissionsPath = path.resolve(__dirname, '../../fixtures/form-submissions.json');
  const raw = fs.readFileSync(submissionsPath, 'utf-8');
  const parsed = JSON.parse(raw) as { submissions?: FormSubmission[] };
  cachedSubmissions = parsed.submissions ?? [];
  return cachedSubmissions;
}

function resolveSubmission(): FormSubmission {
  const submissions = loadFormSubmissions();
  const preferred = submissions.find((submission) => submission.form_type === 'ttc_application');
  return preferred ?? submissions[0] ?? { form_type: 'ttc_application', data: {} };
}

function resolveEmail(context: unknown, submission: FormSubmission): string {
  if (context && typeof context === 'object' && 'current_user' in context) {
    const user = (context as { current_user?: { email?: string } | null }).current_user;
    if (user && typeof user === 'object' && 'email' in user) {
      const email = (user as { email?: string }).email;
      if (typeof email === 'string' && email) {
        return email;
      }
    }
  }
  if (submission.email) {
    return submission.email;
  }
  return 'test.applicant@example.com';
}

When('I upload form data for a specific form instance', async function (this: World) {
  const submission = resolveSubmission();
  const formType = submission.form_type ?? 'ttc_application';
  const formInstance = submission.form_instance ?? 'default';
  const formData = submission.data ?? {};
  const formInstancePageData = submission.form_instance_page_data ?? {};
  const formInstanceDisplay = submission.id ?? submission.ttc_option ?? 'default';

  // Create mock user and upload data
  const user = new MockTTCPortalUser();
  const userEmail = resolveEmail(this, submission);
  user.loadUserData(userEmail);

  // Set form data
  user.setFormData(formType, formInstance, formData, formInstancePageData, formInstanceDisplay);

  // Store the upload details for verification
  userFormContext.lastUpload = {
    formType,
    formInstance,
    formData,
    formInstanceDisplay,
    formInstancePageData,
    user,
  };
});

Then('my form data should be stored for that instance', function () {
  assert.ok(userFormContext.lastUpload, 'Expected uploaded_form_data to be set');

  const { user, formType, formInstance, formData, formInstanceDisplay } = userFormContext.lastUpload;

  // Retrieve the stored data
  const storedData = user.getFormData(formType, formInstance);

  assert.ok(storedData, 'Expected stored data to not be null');
  assert.ok('data' in storedData, 'Expected stored data to have "data" field');

  // Verify the actual form field values were stored
  const storedFormData = storedData.data;
  for (const [key, value] of Object.entries(formData)) {
    assert.strictEqual(
      storedFormData[key],
      value,
      `Expected field ${key} to be ${value}, got ${storedFormData[key]}`
    );
  }

  // Verify metadata fields exist
  assert.ok('is_form_complete' in storedData, 'Expected is_form_complete field');
  assert.ok('is_agreement_accepted' in storedData, 'Expected is_agreement_accepted field');
  assert.ok('is_form_submitted' in storedData, 'Expected is_form_submitted field');
  assert.ok('last_update_datetime' in storedData, 'Expected last_update_datetime field');
  assert.ok('form_instance_display' in storedData, 'Expected form_instance_display field');
  assert.ok('form_instance_page_data' in storedData, 'Expected form_instance_page_data field');

  // Verify form_instance_display matches what was uploaded
  assert.strictEqual(
    storedData.form_instance_display,
    formInstanceDisplay,
    'Expected form_instance_display to match'
  );
});

// TypeScript World interface (basic)
interface World {
  attach?: (attachment: { data: string; media: string }) => void;
}

// Config management context
interface ConfigContext {
  ttcUser?: MockTTCPortalUser;
  userConfig?: Record<string, unknown>;
  lastConfigUpdate?: Record<string, unknown>;
}

const configContext: ConfigContext = {};

When('I request my user configuration', function (this: World) {
  // Get or create ttcUser from configContext
  let ttcUser = configContext.ttcUser;
  if (!ttcUser) {
    // Create a mock user if not already present
    ttcUser = new MockTTCPortalUser();

    // Get email from authContext if available
    const currentUser = authContext.currentUser;
    let email = 'test.applicant@example.com';
    if (currentUser && currentUser.email) {
      email = currentUser.email;
    }

    ttcUser.loadUserData(email);
    configContext.ttcUser = ttcUser;
  }

  configContext.userConfig = ttcUser.getConfig();
});

Then('I should receive my saved configuration', function () {
  const userConfig = configContext.userConfig;
  assert.ok(userConfig, 'No configuration was retrieved');
  assert.ok(typeof userConfig === 'object', 'Configuration should be an object');
  assert.ok(userConfig !== null, 'Configuration should not be null');
});

When('I update my user configuration', function (this: World) {
  // Get or create ttcUser from configContext
  let ttcUser = configContext.ttcUser;
  if (!ttcUser) {
    // Create a mock user if not already present
    ttcUser = new MockTTCPortalUser();

    // Get email from authContext if available
    const currentUser = authContext.currentUser;
    let email = 'test.applicant@example.com';
    if (currentUser && currentUser.email) {
      email = currentUser.email;
    }

    ttcUser.loadUserData(email);
    configContext.ttcUser = ttcUser;
  }

  const testConfig = { i_home_country: 'IN' };
  ttcUser.setConfig(testConfig);
  configContext.lastConfigUpdate = testConfig;
});

Then('my configuration should be saved', function () {
  const ttcUser = configContext.ttcUser;
  assert.ok(ttcUser, 'User should be authenticated');

  const lastUpdate = configContext.lastConfigUpdate;
  assert.ok(lastUpdate, 'Config should have been updated');

  const savedConfig = ttcUser!.getConfig();
  assert.ok(savedConfig, 'Saved config should not be null');

  // Verify the update was persisted
  for (const [key, value] of Object.entries(lastUpdate!)) {
    assert.strictEqual(savedConfig[key], value, `Config key ${key} should be ${value}`);
  }
});

// Get Form Data Context
interface GetFormDataContext {
  savedFormUser?: MockTTCPortalUser;
  savedFormType?: string;
  savedFormInstance?: string;
  savedFormData?: Record<string, unknown>;
  retrievedFormData?: Record<string, unknown>;
}

const getFormDataContext: GetFormDataContext = {};

Given('I have previously saved form data for a form instance', async function () {
  const submission = resolveSubmission();
  const formType = submission.form_type ?? 'ttc_application';
  const formInstance = submission.form_instance ?? 'default';
  const formData = submission.data ?? { i_fname: 'John', i_lname: 'Doe' };
  const formInstancePageData = submission.form_instance_page_data ?? {};
  const formInstanceDisplay = submission.id ?? submission.ttc_option ?? 'default';

  // Create mock user and save data
  const user = new MockTTCPortalUser();
  const userEmail = resolveEmail(this, submission);
  user.loadUserData(userEmail);

  user.setFormData(formType, formInstance, formData, formInstancePageData, formInstanceDisplay);

  // Store in context for later steps
  getFormDataContext.savedFormUser = user;
  getFormDataContext.savedFormType = formType;
  getFormDataContext.savedFormInstance = formInstance;
  getFormDataContext.savedFormData = formData;
});

When('I request that form data', async function (this: World) {
  const user = getFormDataContext.savedFormUser;
  assert.ok(user, 'Expected savedFormUser to be set in context');

  const formType = getFormDataContext.savedFormType ?? 'ttc_application';
  const formInstance = getFormDataContext.savedFormInstance ?? 'default';

  // Call getFormData to retrieve the stored data
  const retrievedData = user.getFormData(formType, formInstance);
  assert.ok(retrievedData, 'Expected getFormData to return data');

  // Extract the 'data' field from the stored form structure
  getFormDataContext.retrievedFormData = retrievedData.data ?? {};
});

Then('I should receive the stored form data', function () {
  const retrieved = getFormDataContext.retrievedFormData;
  const original = getFormDataContext.savedFormData;

  assert.ok(retrieved, 'Expected retrievedFormData to be set');
  assert.ok(original, 'Expected savedFormData to be set');

  // Verify all fields from original data are in retrieved data
  for (const [key, value] of Object.entries(original)) {
    assert.ok(key in retrieved, `Expected key ${key} to be in retrieved data`);
    assert.strictEqual(
      retrieved[key],
      value,
      `Expected ${key} to be ${value}, got ${retrieved[key]}`
    );
  }
});

Given('I have multiple form instances for a form type', function () {
  const user = new MockTTCPortalUser();
  user.loadUserData('test.applicant@example.com');

  // Set up multiple form instances for ttc_application
  user.setFormData(
    'ttc_application',
    'test_us_future',
    { i_fname: 'John', i_lname: 'Doe', i_email: 'john@example.com' },
    { dates: 'Jan 2025', country: 'US' },
    'US TTC - January 2025'
  );

  user.setFormData(
    'ttc_application',
    'test_india_future',
    { i_fname: 'Jane', i_lname: 'Smith', i_email: 'jane@example.com' },
    { dates: 'Feb 2025', country: 'IN' },
    'India TTC - February 2025'
  );

  userFormContext.user = user;
  userFormContext.formType = 'ttc_application';
});

When('I request the list of form instances', function () {
  if (!userFormContext.user) {
    throw new Error('User not initialized');
  }
  userFormContext.formInstances = userFormContext.user.getFormInstances(
    userFormContext.formType as string
  );
});

Then('I should receive the available form instances', function () {
  assert.ok(userFormContext.formInstances, 'Form instances should be defined');
  assert.ok(typeof userFormContext.formInstances === 'object', 'Form instances should be an object');

  const instances = userFormContext.formInstances as Record<string, unknown>;

  // Verify 'default' is excluded
  assert.ok(!('default' in instances), "'default' instance should not be in the results");

  // Verify we have the expected instances
  assert.ok('test_us_future' in instances, 'test_us_future instance should be present');
  assert.ok('test_india_future' in instances, 'test_india_future instance should be present');

  // Verify structure
  const usInstance = instances.test_us_future as Record<string, unknown>;
  const indiaInstance = instances.test_india_future as Record<string, unknown>;

  assert.ok('page_data' in usInstance, 'US instance should have page_data');
  assert.ok('display' in usInstance, 'US instance should have display');
  assert.strictEqual(usInstance.display, 'US TTC - January 2025');

  const usPageData = usInstance.page_data as Record<string, unknown>;
  assert.strictEqual(usPageData.country, 'US');

  assert.ok('page_data' in indiaInstance, 'India instance should have page_data');
  assert.ok('display' in indiaInstance, 'India instance should have display');
  assert.strictEqual(indiaInstance.display, 'India TTC - February 2025');

  const indiaPageData = indiaInstance.page_data as Record<string, unknown>;
  assert.strictEqual(indiaPageData.country, 'IN');
});

File: test/python/steps/user_steps.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import

from behave import given, when, then
import json
import os
import sys

PROJECT_ROOT = os.path.dirname(
    os.path.dirname(os.path.dirname(os.path.dirname(os.path.realpath(__file__))))
)
sys.path.insert(0, PROJECT_ROOT)


class StubUser(object):
    """Stub for Google App Engine users API."""
    def __init__(self, email_addr):
        self._email_addr = email_addr

    def email(self):
        return self._email_addr

    def user_id(self):
        return self._email_addr


class MockTTCPortalUser:
    """Mock TTCPortalUser for testing when GAE dependencies aren't available."""
    def __init__(self):
        self.email = None
        self.form_data = {}
        self.config = {}
        self.is_profile_complete = {}

    def initialize_user(self, user_dict):
        """Initialize user from dictionary."""
        self.email = user_dict.get('email', '')
        self.form_data = {}
        self.config = {}
        self.is_profile_complete = {}

    def set_form_data(self, f_type, f_instance, f_data, f_instance_page_data, f_instance_display):
        """Store form data in the nested dictionary structure."""
        if f_type not in self.form_data:
            self.form_data[f_type] = {}

        # Compute is_form_complete by checking if all fields have non-empty values
        is_form_complete = all(
            bool(v) for v in f_data.values() if v is not None
        )

        self.form_data[f_type][f_instance] = {
            'data': f_data,
            'form_instance_page_data': f_instance_page_data,
            'form_instance_display': f_instance_display,
            'is_agreement_accepted': False,
            'is_form_submitted': False,
            'is_form_complete': is_form_complete,
            'last_update_datetime': '2024-01-01 00:00:00',  # Mock timestamp
        }

        # Also write to 'default' instance if this is a non-default instance
        if f_instance != 'default':
            self.form_data[f_type]['default'] = self.form_data[f_type][f_instance]

    def get_form_data(self, f_type, f_instance):
        """Retrieve stored form data."""
        if f_type in self.form_data and f_instance in self.form_data[f_type]:
            return self.form_data[f_type][f_instance]['data']
        return None

    def load_user_data(self, user_email):
        """Mock load_user_data - initialize user from email."""
        self.email = user_email
        self.initialize_user({'email': user_email})

    def save_user_data(self):
        """Mock save_user_data - no-op for testing."""
        return None

    def get_config(self):
        """Get user configuration."""
        return self.config

    def set_config(self, config_params):
        """Set user configuration parameters."""
        if not isinstance(config_params, dict):
            config_params = json.loads(config_params) if isinstance(config_params, str) else {}

        # Update config with provided params
        for key, value in config_params.items():
            self.config[key] = value

    def get_form_instances(self, f_type):
        """Get all form instances for a form type, excluding 'default'.

        Args:
            f_type: The form type (e.g., 'ttc_application')

        Returns:
            Dictionary of instances with page_data and display info.
        """
        _form_instances = {}
        if f_type in self.form_data:
            for instance_id in self.form_data[f_type]:
                if instance_id != 'default':
                    _form_instances[instance_id] = {
                        'page_data': self.form_data[f_type][instance_id].get(
                            'form_instance_page_data', {}
                        ),
                        'display': self.form_data[f_type][instance_id].get(
                            'form_instance_display', instance_id
                        )
                    }
        return _form_instances


def _get_ttc_portal_user_module():
    """Try to import the real ttc_portal_user module, return None if not available."""
    try:
        import ttc_portal_user
        return ttc_portal_user
    except Exception:
        return None


def _resolve_submission(context):
    """Get form submission from fixtures, preferring ttc_application."""
    submissions = getattr(context, 'fixture_submissions', None) or []
    for submission in submissions:
        if submission.get('form_type') == 'ttc_application':
            return submission
    if submissions:
        return submissions[0]
    return {}


def _resolve_email(context, submission):
    """Get user email from context or submission, with fallback."""
    user = getattr(context, 'current_user', None)
    if isinstance(user, dict):
        email = user.get('email')
        if email:
            return email
    if user is not None and hasattr(user, 'email'):
        try:
            return user.email()
        except Exception:
            pass
    email = submission.get('email')
    if email:
        return email
    return 'test.applicant@example.com'


@when('I upload form data for a specific form instance')
def step_upload_form_data(context):
    """Upload form data using TTCPortalUser.set_form_data() method."""
    submission = _resolve_submission(context)
    form_type = submission.get('form_type', 'ttc_application')
    form_instance = submission.get('form_instance', 'default')
    form_data = submission.get('data', {})
    form_instance_page_data = submission.get('form_instance_page_data', {})
    form_instance_display = (
        submission.get('id')
        or submission.get('ttc_option')
        or submission.get('form_instance_display')
        or 'default'
    )

    # Try to use real TTCPortalUser if available, otherwise use mock
    ttc_portal_user_module = _get_ttc_portal_user_module()
    if ttc_portal_user_module:
        try:
            user = ttc_portal_user_module.TTCPortalUser()
            user_email = _resolve_email(context, submission)
            user.load_user_data(user_email)

            # Call set_form_data with the submission data
            user.set_form_data(
                f_type=form_type,
                f_instance=form_instance,
                f_data=form_data,
                f_instance_page_data=form_instance_page_data,
                f_instance_display=form_instance_display
            )

            # Store the upload details for verification
            context.uploaded_form_data = {
                'form_type': form_type,
                'form_instance': form_instance,
                'form_data': form_data,
                'form_instance_display': form_instance_display,
                'form_instance_page_data': form_instance_page_data,
                'user': user
            }
            return
        except Exception as e:
            # Fall through to mock implementation on error
            pass

    # Use mock implementation when GAE dependencies aren't available
    user = MockTTCPortalUser()
    user_email = _resolve_email(context, submission)
    user.load_user_data(user_email)

    # Call set_form_data with the submission data
    user.set_form_data(
        f_type=form_type,
        f_instance=form_instance,
        f_data=form_data,
        f_instance_page_data=form_instance_page_data,
        f_instance_display=form_instance_display
    )

    # Store the upload details for verification
    context.uploaded_form_data = {
        'form_type': form_type,
        'form_instance': form_instance,
        'form_data': form_data,
        'form_instance_display': form_instance_display,
        'form_instance_page_data': form_instance_page_data,
        'user': user
    }


@then('my form data should be stored for that instance')
def step_form_data_should_be_stored(context):
    """Verify that the form data was stored correctly."""
    uploaded = getattr(context, 'uploaded_form_data', None)
    assert uploaded is not None, 'Expected uploaded_form_data to be set'

    user = uploaded.get('user')
    assert user is not None, 'Expected user object to be set'

    form_type = uploaded.get('form_type')
    form_instance = uploaded.get('form_instance')
    original_data = uploaded.get('form_data')

    # Retrieve the stored data
    stored_data = user.get_form_data(form_type, form_instance)

    assert stored_data is not None, 'Expected stored data to not be None'
    assert 'data' in stored_data, 'Expected stored data to have "data" field'

    # Verify the actual form field values were stored
    stored_form_data = stored_data.get('data', {})
    for key, value in original_data.items():
        assert stored_form_data.get(key) == value, \
            'Expected field {} to be {}, got {}'.format(
                key, value, stored_form_data.get(key)
            )

    # Verify metadata fields exist
    assert 'is_form_complete' in stored_data, 'Expected is_form_complete field'
    assert 'is_agreement_accepted' in stored_data, 'Expected is_agreement_accepted field'
    assert 'is_form_submitted' in stored_data, 'Expected is_form_submitted field'
    assert 'last_update_datetime' in stored_data, 'Expected last_update_datetime field'
    assert 'form_instance_display' in stored_data, 'Expected form_instance_display field'
    assert 'form_instance_page_data' in stored_data, 'Expected form_instance_page_data field'

    # Verify form_instance_display matches what was uploaded
    assert stored_data.get('form_instance_display') == uploaded.get('form_instance_display'), \
        'Expected form_instance_display to match'


@when('I request my user configuration')
def request_user_config(context):
    """Request the user's configuration."""
    # Get or create ttc_user from context
    ttc_user = getattr(context, 'ttc_user', None)
    if not ttc_user:
        # Create a mock user if not already present
        ttc_user = MockTTCPortalUser()
        # Get email from current_user if available
        current_user = getattr(context, 'current_user', None)
        email = 'test.applicant@example.com'
        if current_user:
            if isinstance(current_user, dict):
                email = current_user.get('email', email)
            elif hasattr(current_user, 'email'):
                try:
                    email = current_user.email()
                except Exception:
                    pass
        ttc_user.load_user_data(email)
        context.ttc_user = ttc_user

    context.user_config = ttc_user.get_config()


@then('I should receive my saved configuration')
def should_receive_saved_config(context):
    """Verify that user configuration is received."""
    assert hasattr(context, 'user_config'), "No configuration was retrieved"
    assert isinstance(context.user_config, dict), "Configuration should be a dictionary"
    # For the get scenario, we expect an empty config initially
    assert context.user_config is not None, "Configuration should not be None"


@when('I update my user configuration')
def update_user_config(context):
    """Update the user's configuration with test data."""
    # Get or create ttc_user from context
    ttc_user = getattr(context, 'ttc_user', None)
    if not ttc_user:
        # Create a mock user if not already present
        ttc_user = MockTTCPortalUser()
        # Get email from current_user if available
        current_user = getattr(context, 'current_user', None)
        email = 'test.applicant@example.com'
        if current_user:
            if isinstance(current_user, dict):
                email = current_user.get('email', email)
            elif hasattr(current_user, 'email'):
                try:
                    email = current_user.email()
                except Exception:
                    pass
        ttc_user.load_user_data(email)
        context.ttc_user = ttc_user

    # Sample config update
    test_config = {'i_home_country': 'IN'}
    ttc_user.set_config(test_config)
    context.last_config_update = test_config


@then('my configuration should be saved')
def config_should_be_saved(context):
    """Verify that configuration was saved."""
    ttc_user = getattr(context, 'ttc_user', None)
    assert ttc_user is not None, "User should be authenticated"
    assert hasattr(context, 'last_config_update'), "Config should have been updated"

    saved_config = ttc_user.get_config()
    assert saved_config is not None, "Saved config should not be None"
    # Verify the update was persisted
    for key, value in context.last_config_update.items():
        assert saved_config.get(key) == value, "Config key {} should be {}".format(key, value)


# Get Form Data Steps


@given('I have previously saved form data for a form instance')
def step_given_saved_form_data(context):
    """Set up test context with previously saved form data."""
    # Load form submission from fixtures
    submission = _resolve_submission(context)
    form_type = submission.get('form_type', 'ttc_application')
    form_instance = submission.get('form_instance', 'default')
    form_data = submission.get('data', {'i_fname': 'John', 'i_lname': 'Doe'})
    form_instance_page_data = submission.get('form_instance_page_data', {})
    form_instance_display = (
        submission.get('id')
        or submission.get('ttc_option')
        or 'default'
    )

    # Try to use real TTCPortalUser if available, otherwise use mock
    ttc_portal_user_module = _get_ttc_portal_user_module()
    if ttc_portal_user_module:
        try:
            user = ttc_portal_user_module.TTCPortalUser()
            user_email = _resolve_email(context, submission)
            user.load_user_data(user_email)
            user.set_form_data(
                f_type=form_type,
                f_instance=form_instance,
                f_data=form_data,
                f_instance_page_data=form_instance_page_data,
                f_instance_display=form_instance_display
            )
            context.saved_form_user = user
            context.saved_form_type = form_type
            context.saved_form_instance = form_instance
            context.saved_form_data = form_data
            return
        except Exception:
            pass

    # Use mock implementation
    user = MockTTCPortalUser()
    user_email = _resolve_email(context, submission)
    user.load_user_data(user_email)
    user.set_form_data(
        f_type=form_type,
        f_instance=form_instance,
        f_data=form_data,
        f_instance_page_data=form_instance_page_data,
        f_instance_display=form_instance_display
    )
    context.saved_form_user = user
    context.saved_form_type = form_type
    context.saved_form_instance = form_instance
    context.saved_form_data = form_data


@when('I request that form data')
def step_request_form_data(context):
    """Request the previously saved form data."""
    user = getattr(context, 'saved_form_user', None)
    assert user is not None, 'Expected saved_form_user to be set in context'

    form_type = getattr(context, 'saved_form_type', 'ttc_application')
    form_instance = getattr(context, 'saved_form_instance', 'default')

    # Call get_form_data to retrieve the stored data
    retrieved_data = user.get_form_data(form_type, form_instance)
    context.retrieved_form_data = retrieved_data


@then('I should receive the stored form data')
def step_should_receive_stored_data(context):
    """Verify that the retrieved form data matches what was stored."""
    assert hasattr(context, 'retrieved_form_data'), 'Expected retrieved_form_data to be set'
    assert hasattr(context, 'saved_form_data'), 'Expected saved_form_data to be set'

    retrieved = context.retrieved_form_data
    original = context.saved_form_data

    # Verify all fields from original data are in retrieved data
    for key, value in original.items():
        assert key in retrieved, 'Expected key {} to be in retrieved data'.format(key)
        assert retrieved[key] == value, \
            'Expected {} to be {}, got {}'.format(key, value, retrieved[key])


# Get Form Instances Steps


@given('I have multiple form instances for a form type')
def step_multiple_form_instances(context):
    """Set up a user with multiple form instances for testing."""
    _ttc_user = _get_ttc_portal_user_module()
    if _ttc_user:
        user = _ttc_user.TTCPortalUser()
    else:
        user = MockTTCPortalUser()

    user.load_user_data('test.applicant@example.com')

    # Set up multiple form instances for ttc_application
    user.set_form_data(
        'ttc_application',
        'test_us_future',
        {'i_fname': 'John', 'i_lname': 'Doe', 'i_email': 'john@example.com'},
        {'dates': 'Jan 2025', 'country': 'US'},
        'US TTC - January 2025'
    )

    user.set_form_data(
        'ttc_application',
        'test_india_future',
        {'i_fname': 'Jane', 'i_lname': 'Smith', 'i_email': 'jane@example.com'},
        {'dates': 'Feb 2025', 'country': 'IN'},
        'India TTC - February 2025'
    )

    context.user = user
    context.form_type = 'ttc_application'


@when('I request the list of form instances')
def step_request_form_instances(context):
    """Request the list of form instances for the stored form type."""
    context.form_instances = context.user.get_form_instances(context.form_type)


@then('I should receive the available form instances')
def step_receive_form_instances(context):
    """Verify that form instances are returned correctly."""
    assert isinstance(context.form_instances, dict), \
        "Form instances should be a dictionary"

    # Verify 'default' is excluded
    assert 'default' not in context.form_instances, \
        "'default' instance should not be in the results"

    # Verify we have the expected instances
    assert 'test_us_future' in context.form_instances, \
        "test_us_future instance should be present"
    assert 'test_india_future' in context.form_instances, \
        "test_india_future instance should be present"

    # Verify structure of each instance
    for instance_id, instance_data in context.form_instances.items():
        assert 'page_data' in instance_data, \
            "Instance {} should have 'page_data'".format(instance_id)
        assert 'display' in instance_data, \
            "Instance {} should have 'display'".format(instance_id)

    # Verify specific values
    assert context.form_instances['test_us_future']['display'] == 'US TTC - January 2025'
    assert context.form_instances['test_us_future']['page_data']['country'] == 'US'
    assert context.form_instances['test_india_future']['display'] == 'India TTC - February 2025'
    assert context.form_instances['test_india_future']['page_data']['country'] == 'IN'
